data/tasks.js

```diff
@@ -0,0 +1,217 @@
+    const taskExists = await tasks.findOne({ title: title });
+
+    if (taskExists) {
+      throw new Error(
+        `Task title already exists for the User ${user.first_name}`
+      );
+    }
+    const insertInfo = await tasks.insertOne(newTask);
+
+    if (insertInfo.insertedCount === 0) {
+      throw new Error("Could not add task");
+    }
+
+    const newId = insertInfo.insertedId;
+    await users.updateOne(
+      { _id: new ObjectId(userId) },
+      { $push: { taskIds: newId } }
+    );
+    return await this.getTaskById(newId.toString());
+  },
+
+  async updateTask(id, updatedTask) {
+    utils.checkObjectIdString(id);
+
+    const tasks = await tasksCollection();
+
+    const updatedTaskData = {};
+    updatedTask.title = updatedTask.title.trim();
+    updatedTask.textBody = updatedTask.textBody.trim();
+    updatedTask.tag = updatedTask.tag.trim().toLowerCase();
+
+    if (updatedTask.title) {
+      utils.validateStringInputWithMaxLength(
+        updatedTask.title,
+        "title",
+        constants.stringLimits["title"]
+      );
+      updatedTaskData.title = updatedTask.title;
+    } else {
+      throw new Error("You must provide a title for the task.");
+    }
+
+    if (updatedTask.textBody) {
+      utils.validateStringInputWithMaxLength(
+        updatedTask.textBody,
+        "textBody",
+        constants.stringLimits["textBody"]
+      );
+      updatedTaskData.textBody = updatedTask.textBody;
+    } else {
+      throw new Error("You must provide a textBody for the task.");
+    }
+    if (updatedTask.dateAddedTo) {
+      utils.validateDate(updatedTask.dateAddedTo, "dateAddedTo");
+      updatedTaskData.dateAddedTo = new Date(updatedTask.dateAddedTo);
+    } else {
+      throw new Error("You must provide a dateAddedTo for the task.");
+    }
+
+    if (updatedTask.dateDueOn) {
+      utils.validateDate(updatedTask.dateDueOn, "dateDueOn");
+      updatedTaskData.dateDueOn = new Date(updatedTask.dateDueOn);
+    } else {
+      throw new Error("You must provide a dateDueOn for the task.");
+    }
+
+    if (updatedTask.priority) {
+      utils.validatePriority(updatedTask.priority);
+      updatedTaskData.priority = updatedTask.priority;
+    } else {
+      throw new Error("You must provide a priority for the task.");
+    }
+
+    if (updatedTask.tag) {
+      utils.validateStringInputWithMaxLength(
+        updatedTask.tag,
+        "tag",
+        constants.stringLimits["tag"]
+      );
+      updatedTaskData.tag = updatedTask.tag;
+    } else {
+      throw new Error("You must provide a tag for the task.");
+    }
+
+    if (typeof updatedTask.checked !== "undefined") {
+      utils.validateBooleanInput(updatedTask.checked, "checked");
+      updatedTaskData.checked = updatedTask.checked;
+    } else {
+      throw new Error("You must provide a checked value for the task.");
+    }
+
+    const updateInfo = await tasks.updateOne(
+      { _id: new ObjectId(id) },
+      { $set: updatedTaskData }
```

Comment:
u need to also include the original data like type and dateCreated in this as the set will cause the document to be overwritten

suggested code before passing would be
const oldTask = await this.get(taskId)
updatedTaskData.type = oldTask.type
updatedTaskData.dateCreated = oldTask.dateCreated

similarly for any fields you have missed